cherrypy/lib/sessions.py

```diff
@@ -404,18 +404,28 @@ def clean_up(self):
         """Clean up expired sessions."""
 
         now = self.now()
-        for _id, (data, expiration_time) in list(self.cache.items()):
-            if expiration_time <= now:
-                try:
-                    del self.cache[_id]
-                except KeyError:
-                    pass
-                try:
-                    if self.locks[_id].acquire(blocking=False):
-                        lock = self.locks.pop(_id)
-                        lock.release()
-                except KeyError:
-                    pass
+        try:
+            for _id, (data, expiration_time) in list(self.cache.items()):
```

Comment:
Honestly, I'm not sure about the try/except. It just hides the problem.

I think that the initial purpose of `list()` was to copy the contents so that the for-loop would iterate over an immutable iterable. But it seems like in the case you describe the copying process (new list creation) *lazily* iterates over the dict entries (`.items()` returns an iterator). And that's where it fails, not in the for-loop as you might've thought.

That said, the proper patch following your way would be to wrap just `list(self.cache.items())` with try/except. Yet, swallowing a RuntimeError like that doesn't feel right.

I think that the ultimate solution would be to actually copy the dict which, I think, is an atomic operation. So I'd apply this instead of wrapping everything with try/except:
```suggestion
            for _id, (data, expiration_time) in self.cache.copy().items():
```

Reply:
any suggestions on how to test this alternative?

Reply:
I've done some research and I'm pretty sure we can rely on this:
```python
In [1]: d1 = {1: 2}                                                                                                                     

In [2]: import dis                                                                                                                      

In [3]: dis.dis('d2 = d1.copy()')                                                                                                       
  1           0 LOAD_NAME                0 (d1)
              2 LOAD_METHOD              1 (copy)
              4 CALL_METHOD              0
              6 STORE_NAME               2 (d2)
              8 LOAD_CONST               0 (None)
             10 RETURN_VALUE

In [4]: dis.dis('d3 = dict(d1)')                                                                                                        
  1           0 LOAD_NAME                0 (dict)
              2 LOAD_NAME                1 (d1)
              4 CALL_FUNCTION            1
              6 STORE_NAME               2 (d3)
              8 LOAD_CONST               0 (None)
             10 RETURN_VALUE

In [5]: dis.dis('list(d1.items())')                                                                                                     
  1           0 LOAD_NAME                0 (list)
              2 LOAD_NAME                1 (d1)
              4 LOAD_METHOD              2 (items)
              6 CALL_METHOD              0
              8 CALL_FUNCTION            1
             10 RETURN_VALUE
```

OTOH if you want to test you could deploy the following in production and see if the error is still happening:
```suggestion
            try:
                cache_items_copy = self.cache.copy().items()
            except RuntimeError as re:
                cherrypy.log(f'Runtime Error happened while copying the cache entries: {re}')
                cherrypy.log('Ref: https://github.com/cherrypy/cherrypy/pull/1804')
                return
            for _id, (data, expiration_time) in cache_items_copy:
```
without that outer try/except.

Alternatively, you could try reproducing this using `ab`/`Locust' or similar.

Reply:
I tried to reproduce the error in some test code without success. I will implement your suggestion in production and see if the runtimeerror occurs with cache.copy().items()